gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageImpl.java

```diff
@@ -964,63 +990,82 @@ public void onFailure(GoogleJsonError jsonError, HttpHeaders responseHeaders) {
+      StorageResourceId destination = entry.getValue();
+      String srcBucketName = source.getBucketName();
+      String dstBucketName = destination.getBucketName();
+      // Avoid copy across locations or storage classes.
+      if (!srcBucketName.equals(dstBucketName)) {
+        StorageResourceId srcBucketResourceId = new StorageResourceId(srcBucketName);
+        GoogleCloudStorageItemInfo srcBucketInfo =
+            getGoogleCloudStorageItemInfo(gcsImpl, bucketInfoCache, srcBucketResourceId);
+        if (!srcBucketInfo.exists()) {
+          throw new FileNotFoundException("Bucket not found: " + srcBucketName);
+        }
+
+        StorageResourceId dstBucketResourceId = new StorageResourceId(dstBucketName);
+        GoogleCloudStorageItemInfo dstBucketInfo =
+            getGoogleCloudStorageItemInfo(gcsImpl, bucketInfoCache, dstBucketResourceId);
+        if (!dstBucketInfo.exists()) {
+          throw new FileNotFoundException("Bucket not found: " + dstBucketName);
         }
 
-        if (!srcBucketInfo.getStorageClass().equals(dstBucketInfo.getStorageClass())) {
-          throw new UnsupportedOperationException(
-              "This operation is not supported across two different storage classes.");
+        if (!gcsImpl.getOptions().isCopyWithRewriteEnabled()) {
+          if (!srcBucketInfo.getLocation().equals(dstBucketInfo.getLocation())) {
+            throw new UnsupportedOperationException(
+                "This operation is not supported across two different storage locations.");
+          }
+
+          if (!srcBucketInfo.getStorageClass().equals(dstBucketInfo.getStorageClass())) {
+            throw new UnsupportedOperationException(
+                "This operation is not supported across two different storage classes.");
+          }
         }
       }
-    }
-    for (int i = 0; i < srcObjectNames.size(); i++) {
-      Preconditions.checkArgument(
-          !isNullOrEmpty(srcObjectNames.get(i)), "srcObjectName must not be null or empty");
-      Preconditions.checkArgument(
-          !isNullOrEmpty(dstObjectNames.get(i)), "dstObjectName must not be null or empty");
+      checkArgument(
+          !isNullOrEmpty(source.getObjectName()), "srcObjectName must not be null or empty");
+      checkArgument(
+          !isNullOrEmpty(destination.getObjectName()), "dstObjectName must not be null or empty");
       if (srcBucketName.equals(dstBucketName)
-          && srcObjectNames.get(i).equals(dstObjectNames.get(i))) {
+          && source.getObjectName().equals(destination.getObjectName())) {
         throw new IllegalArgumentException(
             String.format(
                 "Copy destination must be different from source for %s.",
-                StringPaths.fromComponents(srcBucketName, srcObjectNames.get(i))));
+                StringPaths.fromComponents(srcBucketName, source.getObjectName())));
       }
     }
   }
 
+  private static GoogleCloudStorageItemInfo getGoogleCloudStorageItemInfo(
```

Comment:
You can replace this whole method with `Map.computeIfAbsent` method.

Reply:
`getItemInfo` throws exception, `Map.computeIfAbsent` accepts a lambda expression, which does not allow an unhandled exception, any suggestions ?

Reply:
I think you can catch an exception inside the lambda, ideally we would use `TunnelException`, but it is not open source yet: https://github.com/google/guava/issues/2828#issuecomment-304187823

Reply:
Discussed offline, `RuntimeException` will not work here, but we still can try to make this method a little more optima and concise:

```java
    GoogleCloudStorageItemInfo storageItemInfo = bucketInfoCache.get(resourceId);
    if (storageItemInfo != null) {
      return storageItemInfo;
    }
    storageItemInfo = gcsImpl.getItemInfo(resourceId);
    bucketInfoCache.put(resourceId, storageItemInfo);
    return storageItemInfo;
```